components/core/src/clp/string_utils/string_utils.cpp

```diff
@@ -182,114 +182,157 @@ bool wildcard_match_unsafe(string_view tame, string_view wild, bool case_sensiti
+ * `wild`, the entire wildcard match fails.
  *
- * Thus, the algorithm:
- * 1. searches for the start of one of these groups in wild,
- * 2. searches for a group in tame starting with the same character, and then
- * 3. checks if the two match. If not, the search repeats with the next group in
- *    tame.
+ * NOTE:
+ * - Since the caller guarantees that there are no consecutive '*', we don't need to handle the
+ *   case where a group in `wild` is empty.
+ * - Since the caller guarantees that every '\' is followed by a character, we can advance passed
+ *   '\' without doing a subsequent bounds check.
+ * - The second part of this method could be rewritten in the following form:
+ *
+ *   ```
+ *   while(true) {
+ *     if (false == advance_tame_to_next_match(...)) return false;
+ *
+ *     while (true) {
+ *       // Advance iterators
+ *       // If we reached the end of `tame` before the end of `wild`, break
+ *       // If we see a '*' in `wild`, break
+ *       // If we see a mismatch, break
+ *     }
+ *   }
+ *   ```
+ *
+ *   However, this form is ~2% slower.
  */
 bool wildcard_match_unsafe_case_sensitive(string_view tame, string_view wild) {
-    auto const tame_length = tame.length();
-    auto const wild_length = wild.length();
-    char const* tame_current = tame.data();
-    char const* wild_current = wild.data();
-    char const* tame_bookmark = nullptr;
-    char const* wild_bookmark = nullptr;
-    char const* tame_end = tame_current + tame_length;
-    char const* wild_end = wild_current + wild_length;
+    auto tame_it = tame.cbegin();
+    auto wild_it = wild.cbegin();
+    auto const tame_end_it = tame.cend();
+    auto const wild_end_it = wild.cend();
+    string_view::const_iterator tame_bookmark_it{};
+    string_view::const_iterator wild_bookmark_it{};
 
-    // Handle wild or tame being empty
-    if (0 == wild_length) {
-        return 0 == tame_length;
-    } else {
-        if (0 == tame_length) {
-            return "*" == wild;
+    // Handle `tame` or `wild` being empty
+    if (wild.empty()) {
+        return tame.empty();
+    }
+    if (tame.empty()) {
+        return "*" == wild;
+    }
+
+    // Match `tame` against `wild` against until we reach the first '*' in `wild` or they no longer
+    // match
+    while (true) {
+        auto w = *wild_it;
+        if ('*' == w) {
+            break;
+        }
+        if ('?' != w) {
+            // Handle escaped characters
+            if ('\\' == w) {
+                // Safe without a bounds check
+                ++wild_it;
+                w = *wild_it;
+            }
+
+            // Handle a mismatch
+            if (w != *tame_it) {
+                return false;
+            }
+        }
+
+        ++tame_it;
+        ++wild_it;
+
+        // Handle boundary conditions
+        if (tame_end_it == tame_it) {
+            return (wild_end_it == wild_it) || (wild_end_it == wild_it + 1 && '*' == *wild_it);
```

Comment:
This line is repeated in the next for loop. How about we make a helper, sth like `is_wild_reaching_end_or_trailing_star`? (maybe we can discuss to come up with a better name)